### PR TITLE
DeepSeek V4 HCA: functional prefill

### PR DESCRIPTION
**Pull request by jmitrovicTT**
### PR Category
Feature

### Issue
Compressor module DS V4 #47211

### Summary
DeepSeek-V4 Heavily Compressed Attention (HCA) for prefill.
Two modules in `tt/mla/heavily_compressed_attention.py`: `TtHCACompressor` (windowed KV
compression) and `TtHCA` (the attention block). Both are mesh-aware — tokens sharded over
the SP axis, heads/hidden over TP. Chunked prefill carries `TtHCAState` across calls; every
tensor shape is fixed for the whole prefill.

**Limitations** `actual seq len` must be divisible by 128

**Accuracy** — PCC vs the torch reference: 0.9925 flash / 0.9816 pro on the last of 11 chunks (drift
accumulates through the compressed cache).

**Perf** — one 5120-token chunk on the 8x4 galaxy: **4.46 ms** flash, **8.13 ms** pro.

| | flash µs | % | pro µs | % |
|---|---|---|---|---|
| SDPA | 1,779 | 39.9 | 4,112 | 50.6 |
| layout / TM | 1,016 | 22.8 | 1,483 | 18.2 |
| CCL | 824 | 18.5 | 980 | 12.0 |
| Matmul | 554 | 12.4 | 1,190 | 14.6 |
| other | 283 | 6.3 | 367 | 4.5 |

Full 56,320-token prefill (11 × 5120): **49.0 ms** flash, **89.4 ms** pro. Cost is flat across
chunks — `Sk` is fixed at 5,792 for the whole prefill (carry 128 + chunk 5,120 + cache 544), because
the sliding window is fixed and the whole compressed cache is read every chunk.

**Tests** — `tests/pcc/test_ttnn_hca.py` covers the compressor, single-shot forward, chunked prefill and the 56K run, all driving the public `forward`. `tests/perf/test_ttnn_hca_perf.py` is the device-perf gate. The 3-chunk chunked scenario (~2 min) is added to Blaze Models Prefill as `hca_chunked`.

### Notes for reviewers
- [x] [HCA: Blaze Models Prefill tests](https://github.com/tenstorrent/tt-metal/actions/runs/32062161335/job/95488460653) passed✅ (3 minutes pcc + 3 minutes perf)
- HCA roofline analysis [table_link](https://docs.google.com/spreadsheets/d/1kOsNPe4sfXvi3NNChXN0wThTjKSrbmf3AgpR4eBKeM4/edit?gid=506570920#gid=506570920) showing the best setup is SP8 and TP4

### Stacked Perf report 2 chunks:

|HCA V4 Flash|HCA V4 PRO|
|--------------|------------|
|hca_2chunk_stacked_flash|hca_2chunk_stacked_pro|

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:jmitrovic/v4_compressor)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:jmitrovic/v4_compressor)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:jmitrovic/v4_compressor)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:jmitrovic/v4_compressor)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:jmitrovic/v4_compressor)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:jmitrovic/v4_compressor)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:jmitrovic/v4_compressor)

### CODE REVIEW

File: models/demos/deepseek_v3_d_p/tests/pcc/test_ttnn_hca.py [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+        return self.layers[layer_idx].update(key_states, value_states)
+
+
+def _sliding_mask(q_pos, k_pos, sliding_window):
+    i, j = q_pos.view(-1, 1), k_pos.view(1, -1)
+    allowed = (j <= i) & (i - j < sliding_window)
+    return torch.zeros(i.shape[0], j.shape[1]).masked_fill(~allowed, float("-inf"))
+
+
+@pytest.mark.timeout(0)
```
[the remark this answers is not in the document]
Comment (jmitrovicTT) [reply models/demos/deepseek_v3_d_p/tests/pcc/test_ttnn_hca.py:507]: Intentional, and consistent with the rest of this suite — `timeout(0)` appears 69 times across models/ and tests/

**Review by mtairum — APPROVED** [review]

File: models/demos/deepseek_v3_d_p/tests/perf/test_ttnn_hca_perf.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        command=f"pytest {_TEST_PATH} -k '{variant} and chunk5120-full and 8x4'",
+        expected_device_perf_ns_per_iteration=expected_ns,
+        subdir=f"deepseek_v4_hca_block_{variant}",
+        model_name=f"deepseek_v4_hca_{variant}_glx_8x4",
+        num_iterations=1,
+        batch_size=1,
+        margin=margin,
+        between_signposts=("HCA_START", "HCA_END"),
+        comments=f"chunk5120_x2_hca_chunked_prefill_{variant}_glx_8x4_ground_truth",
+    )
```
Comment (pavlepopovic) [inline models/demos/deepseek_v3_d_p/tests/perf/test_ttnn_hca_perf.py:95]: are we adding this to CI ? I suggest so, otherwise no value in this test ?

**Review by pavlepopovic — APPROVED** [review]

**Review by roseli-TT — APPROVED** [review]
